### PR TITLE
Add support for profiling with Scalene

### PR DESCRIPTION
**Pull request by chuckwondo**
Added a new input (`scalene_args`) to allow users to run the algorithm via [Scalene](https://github.com/plasma-umass/scalene) to produce a profiling (mem+cpu) report. The input value is simply a string of command-line options to pass to `scalene` to launch the algorithm within the existing `subset.sh` run script.

In addition, added/modified the following to address various inconveniences along the way:

- Combined/simplified conda lock usage to significantly speed up the `build.sh` script that installs deps into the conda environment
- Added `bin/generate-lock-file.sh` script for conveniently regenerating the conda lock file after changes are made to either `environment.yml` or `environment-dev.yml`
- Updated `algorithm_config.yaml` to match new structure/naming implemented on the MAAP API side
- Added convenience scripts for managing algorithm registration, all of which use the updated `algorithm_config.yaml` file:
  - `bin/describe-algorithm.sh`
  - `bin/register-algorithm.sh`
  - `bin/delete-algorithm.sh`
- Updated `docs/MAAP_USAGE.md` accordingly

Here are 2 example profile outputs in the ADE, both using the same algo inputs, but one is a "reduced" profile:

- `~/shared-buckets/dschuck/gedi-subset/profile.html`
- `~/shared-buckets/dschuck/gedi-subset/profile-reduced.html`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,21 +5,30 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog], and this project adheres to
 [Semantic Versioning].
 
-## [0.6.2] - 2023-12-05
```
Comment (wildintellect) [inline CHANGELOG.md:8]: Interesting, did you manually drop the dates from the release versions? or was that an auto-formatting? isn't against semantic versioning to included - because they seem useful.

File: bin/build-dps.sh
Comment (wildintellect) [inline bin/build-dps.sh:1]: Is this file necessary, or can we simplify to just the build.sh?

File: bin/conda-prefix.sh
Comment (wildintellect) [inline bin/conda-prefix.sh:1]: Need to Document what this file's purpose is, comment at top would suffice.

File: bin/delete-algorithm.sh
Comment (wildintellect) [inline bin/delete-algorithm.sh:1]: What's the expected method for executing this script and it's relatives?

File: CONTRIBUTING.md
Comment (wildintellect) [inline CONTRIBUTING.md:1]: TODO: Open a ticket about Documenting how to use this code outside of MAAP.

File: docs/MAAP_USAGE.md [hunk 2]
```diff
@@ -145,14 +132,41 @@ To run a GEDI subsetting DPS job, you must supply the following inputs.
...
+  - Unspecified -> `myaoi.gpkg`, where `myaoi.geojson` is the name of the AOI
+    file
   - `myoutput` -> `myoutput.gpkg`
   - `myoutput.gpkg` -> `myoutput.gpkg`
   - `myoutput.h5` -> `myoutput.gpkg`
   - `mypath/myoutput` -> `mypath/myoutput.gpkg`
 
   _New in version 0.6.0_
 
+- `scalene_args` (_optional_): Arguments to pass to [Scalene] for performance
```
Comment (wildintellect) [inline docs/MAAP_USAGE.md:144]: Suggested rewrite for this section to emphasize what Science users need to know first, developer and more advanced information later.

```
- `scalene_args` (_optional_): Arguments to pass to [Scalene] for performance
  profiling. Normal usage should leave this argument blank.
  Fill this in if you want to collect performance metrics (i.e. CPU and RAM usage)
  The recommended value for this input is  `--reduced-profile` (see below for more advanced usage).
  When used you will find `profile.html` in your algorithm output folder.

  When this input is supplied, the algorithm will be run via the `scalene` command, and the value of this input will be passed as arguments to
  the command.  For a list of the available command-line options, see
  <https://github.com/plasma-umass/scalene?tab=readme-ov-file#scalene>.

  Starting with `--reduced-profile` produces a relatively brief report that may aid in more quickly identifying hotspots than a fully profile would.  However, to produce a full profile where you want to use all of Scalene's default values, you must supply _some_ value for this input, so the simplest valid Scalene option is `--on`.  Otherwise, as mentioned above, when no value is supplied for this input, Scalene will not be used at all.

  > **Note:** that since no browser is available in DPS, when any value is supplied for this input, the `--no-browser` option will be included to prevent Scalene from attempting to open a browser.  However, the `--web`
  option will also be included, which will produce HTML output to a file named `profile.html`.
```

**Review by wildintellect — COMMENTED** [review]
I made some suggestions and asked a few clarifying questions. Nothing is a blocker that I can see, but I would like the readability for regular users addressed before approving.

File: CHANGELOG.md (hunk 1 above)
Comment (chuckwondo) [reply CHANGELOG.md:8]: I removed them from the file for 2 reasons: (1) each is a guess based on when we think the actual release will be made (so it is possible to be wrong), and (2) they are redundant because github automatically adds the release date to the [releases page](https://github.com/MAAP-Project/gedi-subsetter/releases)

File: bin/build-dps.sh
Comment (chuckwondo) [reply bin/build-dps.sh:1]: It _might_ be unnecessary, but I haven't confirmed. _If_ the build command in the algorithm config can actually include arguments, then yes, this becomes unnecessary. I'm going to see if this is the case.

File: bin/conda-prefix.sh
Comment (chuckwondo) [reply bin/conda-prefix.sh:1]: Will do.

File: bin/delete-algorithm.sh
Comment (chuckwondo) [reply bin/delete-algorithm.sh:1]: The delete-algorithm and register-algorithm scripts are referenced in CONTRIBUTING.md. The describe-algorithm script is not described because it is not really intended to be used directly. The other 2 scripts call it.

File: CONTRIBUTING.md
Comment (chuckwondo) [reply CONTRIBUTING.md:1]: I don't think it can be used outside of MAAP. Wouldn't we need to do some refactoring first?

File: docs/MAAP_USAGE.md (hunk 2 above)
Comment (chuckwondo) [reply docs/MAAP_USAGE.md:144]: Will do.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,127 +1,340 @@
...
+requires you to recreate any such conda environments that you wish to use after
+a workspace restart.  To prevent this from occurring, you must provide a
+location that persists across restarts via the `--prefix` option with the
+`bin/build.sh` script, as described below.  _However_, taking this approach
+means that creation of the conda environment may take significantly more time
+(possibly 10s of minutes) because the persistent location is an AWS S3 bucket.
+
+If you're not working in the ADE, or you are, but aren't bothered by having to
+rerun the following command every time you want to use the `gedi_subset` after
+a workspace restart, you may create the conda environment simply as follows:
```
[the remark this answers is not in the document]
Comment (chuckwondo) [reply CONTRIBUTING.md:66]: Ah, good find. I didn't notice that part of the MAAP docs, so I'll link to it.

[the remark this answers is not in the document]
Comment (wildintellect) [reply CONTRIBUTING.md:66]: Also the `root` of `projects` is EFS backed not S3. `conda` environments should not be put in any path labelled `bucket`

File: CONTRIBUTING.md
Comment (wildintellect) [reply CONTRIBUTING.md:1]: @chuckwondo yes, exactly - we can have a document that lists the known things that currently prevent use outside of MAAP, really should just be any used of `maap-py`

File: CHANGELOG.md (hunk 1 above)
Comment (wildintellect) [reply CHANGELOG.md:8]: But that would require a user to go dig into the Github releases and cross reference. It's much nicer to just have it in the log for humans to read.

File: CONTRIBUTING.md
Comment (chuckwondo) [reply CONTRIBUTING.md:1]: ok, just wanted to make sure I was thinking straight -- I'll add a ticket to this repo for this, but should I also add one elsewhere for better visibility?

Comment (wildintellect) [reply CONTRIBUTING.md:1]: It's fine on this repo, we'll track it on the MAAP project board.

File: CONTRIBUTING.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (chuckwondo) [reply CONTRIBUTING.md:66]: Makes sense. Although, even EFS is quite slow, relatively speaking. It took about 20 minutes to install deps under `/projects/envs` vs. about 2-3 minutes under `/opt/conda/envs`.

[the remark this answers is not in the document]
Comment (wildintellect) [reply CONTRIBUTING.md:66]: We should add that to the docs, some users would want to know that - especially since `mamba` is the default now.

File: CONTRIBUTING.md
Comment (chuckwondo) [reply CONTRIBUTING.md:1]: See https://github.com/MAAP-Project/gedi-subsetter/issues/61

File: CHANGELOG.md [hunk 4]
```diff
@@ -42,8 +51,7 @@ The format is based on [Keep a Changelog], and this project adheres to
 ### Changed
 
 - Upgraded Python to version 3.11 to take advantage of the addition of
-  [fine-grained error locations in tracebacks](https://docs.python.org/3/whatsnew/3.11.html#whatsnew311-pep657)
-  to help with debugging errors.
+  [fine-grained error locations in tracebacks] to help with debugging errors.
```
Comment (wildintellect) [inline CHANGELOG.md:54]: Accidentally dropped the link?

Comment (chuckwondo) [reply CHANGELOG.md:54]: I moved it to the bottom of the file.

Comment (wildintellect) [reply CHANGELOG.md:54]: but you still have brackets on this line?

Comment (chuckwondo) [reply CHANGELOG.md:54]: yes, that's how it works: I can use `[X]` within the body of the doc, then put the following at the bottom of the doc so that it does not clutter the main text:

```
[X]:
    https://relevant-url
```

Comment (chuckwondo) [reply CHANGELOG.md:54]: See https://riptutorial.com/markdown/example/2215/reference-link for the various forms

**Review by wildintellect — APPROVED** [review]
